### ISSUE TITLE
bug/feat: errors querying remote parquet with Ibis via DuckDB backend

### ISSUE THREAD

**Issue by alexkyllo** [ibis-project/ibis#4481]
Hello,

I've been experimenting with Ibis as a way to query Parquet files in cloud storage via the DuckDB backend, but have been running into errors like `Python Object module not suitable for replacement scans` and `ParseError: expected one of 'EOF', '[' at 0:9`. Code sample to reproduce below, the cloud files are public access so anyone can query them.

I also tried to do something similar via the DataFusion backend but was unable to figure out how to connect to a cloud storage filesystem via Ibis. (Does it only work with local parquet files?)

Is this a bug? Should I be able to query DuckDB views over Parquet files with Ibis? If not, what is the recommended way to query parquet files in cloud storage with Ibis? Could someone please write a code sample?

```python
import adlfs
import duckdb
import ibis
from pyarrow import dataset

# Example 1:
fs = adlfs.AzureBlobFileSystem(account_name="azureopendatastorage", sas_token="", container_name="mlsamples")
dt = dataset.dataset("mlsamples/diabetes", filesystem=fs)
conn = duckdb.connect("diabetes.db")
conn.register("diabetes", dt)
conn.view("diabetes").aggregate("SEX, count(*) AS ct").df()
# works

ibis_conn = ibis.duckdb.connect("diabetes.db")
ibis_conn.raw_sql("create view diabetes as select * from dt")
diabetes = ibis_conn.table("diabetes")
diabetes.group_by(diabetes.SEX).count().execute()
# RuntimeError: Python Object module not suitable for replacement scans

# Example 2:
fs = adlfs.AzureBlobFileSystem(account_name="azureopendatastorage", sas_token="", container_name="holidaydatacontainer")
dt = dataset.dataset("holidaydatacontainer/Processed", filesystem=fs)
conn = duckdb.connect("holidays.db")
conn.register("holidays", dt)
conn.view("holidays").aggregate("CountryOrRegion, count(*) AS ct").df()
# works

ibis_conn = ibis.duckdb.connect("holidays.db")
ibis_conn.raw_sql("create view holidays as select * from dt")
holidays = ibis_conn.table("holidays")
# ParseError: expected one of 'EOF', '[' at 0:9

**Comment by Mause** [comment]
It looks like the ParseError at least, is due to ibis not supporting `TIMESTAMP_NS` being returned by duckdb

The RuntimeError/Python Object error is due to a name collision in duckdb's "magic" variable frame lookup, though I'm not sure exactly where

**Comment by cpcloud** [comment]
@alexkyllo Thanks for the issue, great writeup!

And thanks @Mause for the diagnosis.

Regarding the `RuntimeError`:

`dt` is probably colliding with a common internal ibis alias: we write `import ibis.expr.datatypes as dt` in many places in the ibis codebase. I'm not sure what to do about that other than tell you to rename `dt` to something else :) since looking at stack frames for variables is a fraught endeavor. It's nice when it works, and hellish when it doesn't.

Regarding the `ParseError`:

This is a bug, I don't think we were aware that `TIMESTAMP_NS` was a valid type. I'll try to come up with a reproducer for this and put up a PR to fix it!

**Comment by Mause** [comment]
Reproducing the `TIMESTAMP_NS` ended up being fairly straightforward:

```python
ibis_conn = ibis.duckdb.connect(":memory:")
ibis_conn.raw_sql("create view holidays as select '2020-01-01T00:00:00Z'::TIMESTAMP_NS")
holidays = ibis_conn.table("holidays")
```

regarding the RuntimeError, I'm going to make some changes upstream in DuckDB to at the very least improve this error message, and possibly add some sort of frame filtering mechanism. Food for thought

**Comment by alexkyllo** [comment]
Thank you both very much! Confirming the first ("diabetes") code sample works if I change the variable name to something other than `dt`.

I also forgot to ask: Is it possible to make the second line below work without having to call the first line with `raw_sql` to create the view? Does Ibis have, or could it add, some way to more directly register a [PyArrow Dataset](https://arrow.apache.org/docs/python/dataset.html) as a selectable object in a DuckDB backend?

```python
ibis_conn.raw_sql("create view diabetes as select * from dt")
diabetes = ibis_conn.table("diabetes")
```

This way is a little awkward, especially because I also need to check if the view exists before creating it--if it does, I get a `RuntimeError: Catalog Error: View with name "diabetes" already exists!`

**Comment by jcrist** [comment]
In #4492 we add ibis support for registering pandas/pyarrow objects as tables with duckdb. Once that's in, you should be able to do something like:

```python
# Create a new ibis duckdb connection
con = ibis.duckdb.connect("diabetes.db")
# Create a pyarrow dataset, and register it with duckdb through ibis
fs = adlfs.AzureBlobFileSystem(account_name="azureopendatastorage", sas_token="", container_name="mlsamples")
dt = dataset.dataset("mlsamples/diabetes", filesystem=fs)
t = con.register(dt, "diabetes")
# run whatever queries you want on `t`
t.head().execute()
```

**Comment by cpcloud** [comment]
@alexkyllo Thanks for creating this issue! In 3.2.0 (coming out today or tomorrow) the `TIMESTAMP_NS` bug will be fixed (by #4495) and you should be able to use various pyarrow objects with the DuckDB backend (from #4492).
